### PR TITLE
Fix gather op bug

### PR DESCRIPTION
**Pull request by chengduoZH**
Related issue https://github.com/PaddlePaddle/Paddle/issues/19085.
The return of `DDim::size()` is the rank info, this is to say, if the `tensor` is empty, the `tensor.dims().size()` is 1.

https://github.com/PaddlePaddle/Paddle/blob/0019eb376ac83fda927025c1c5c3d8cd54fe4b66/paddle/fluid/framework/ddim.h#L61-L62

### CODE REVIEW

File: paddle/fluid/operators/gather.cu.h [hunk 1]
```diff
@@ -51,7 +51,7 @@ void GPUGather(const platform::DeviceContext& ctx, const Tensor& src,
                const Tensor& index, Tensor* output) {
   // PADDLE_ENFORCE(platform::is_gpu_place(place));
   // check index of shape 1-D
-  PADDLE_ENFORCE(index.dims().size() == 1 ||
+  PADDLE_ENFORCE((index.dims().size() == 1 && index.dims()[0] > 0) ||
                  (index.dims().size() == 2 && index.dims()[1] == 1));
```
Comment (luotao1) [inline paddle/fluid/operators/gather.cu.h]:
```
if (index.dims().size() == 1)
  PADDLE_ENFORCE_GT(index.dims()[0], 0, "error_msg");
else if (index.dims().size() == 2)
  PADDLE_ENFORCE_EQ(index.dims()[1], 1, "error_msg");
```

File: paddle/fluid/operators/gather_op.cu [hunk 2]
```diff
@@ -41,6 +41,7 @@ class GatherOpCUDAKernel : public framework::OpKernel<T> {
         paddle::framework::DataTypeToString(index_type),
         paddle::framework::DataTypeToString(framework::proto::VarType::INT32),
         paddle::framework::DataTypeToString(framework::proto::VarType::INT64));
+    PADDLE_ENFORCE_GT(index->numel(), 0, "The index is empty.");
```
Comment (luotao1) [inline paddle/fluid/operators/gather_op.cu]: The error msg is too simple. The index of gather_op should not be empty?

File: paddle/fluid/operators/gather.cu.h (hunk 1 above)
Comment (chengduoZH) [reply paddle/fluid/operators/gather.cu.h]: Done

File: paddle/fluid/operators/gather_op.cu (hunk 2 above)
Comment (chengduoZH) [reply paddle/fluid/operators/gather_op.cu]: Done

File: paddle/fluid/operators/gather.cu.h [hunk 3]
```diff
@@ -51,8 +51,13 @@ void GPUGather(const platform::DeviceContext& ctx, const Tensor& src,
...
+  if (index.dims().size() == 1) {
+    PADDLE_ENFORCE_GT(index.dims()[0], 0, "The index should not be empty.");
+  } else if (index.dims().size() == 2) {
+    PADDLE_ENFORCE_EQ(index.dims()[1], 1, "The index should be 1-D.");
+  }
```
Comment (luotao1) [inline paddle/fluid/operators/gather.cu.h:61]: 1. 有了56-60行，52，54，55行可以去掉了。
2. 57行，59行的报错信息需要增强。举例：
```
The index of gather_op should not be empty when index.dims().size() == 1.
The index of gather_op should be 1-D when index.dims().size() == 2.
```
1-D是1还是1维？

**Review by luotao1 — COMMENTED** [review]
这两个文件中其他PADDLE_ENFORCE也需要改掉，不然CI过不了。

File: paddle/fluid/operators/gather.cu.h [hunk 4]
```diff
@@ -49,10 +49,16 @@ __global__ void GatherCUDAKernel(const T* params, const IndexT* indices,
...
 void GPUGather(const platform::DeviceContext& ctx, const Tensor& src,
                const Tensor& index, Tensor* output) {
-  // PADDLE_ENFORCE(platform::is_gpu_place(place));
   // check index of shape 1-D
-  PADDLE_ENFORCE(index.dims().size() == 1 ||
-                 (index.dims().size() == 2 && index.dims()[1] == 1));
+  if (index.dims().size() == 1) {
+    PADDLE_ENFORCE_GT(index.dims()[0], 0,
+                      "The index of gather_op should not be empty when "
+                      "index.dims().size() == 1.");
```
Comment (chenwhql) [inline paddle/fluid/operators/gather.cu.h]: I think `index.dims().size() == 1` isn't a good description, maybe `when the dimension size of index equals 1.` is better

File: paddle/fluid/operators/gather.cu.h [hunk 5]
```diff
@@ -49,10 +49,16 @@ __global__ void GatherCUDAKernel(const T* params, const IndexT* indices,
...
+  } else if (index.dims().size() == 2) {
+    PADDLE_ENFORCE_EQ(
+        index.dims()[1], 1,
+        "If index.dims().size() is 2, the seconde dim should be 1.");
```
Comment (chenwhql) [inline paddle/fluid/operators/gather.cu.h]: `seconde` -> `second`, maybe the whole sentence can be polish to `If the dimension size of index is 2, the second dimension should be equal to 2.`

**Review by chenwhql — CHANGES_REQUESTED** [review]

**Review by chenwhql — APPROVED** [review]
LGTM

**Pull request merged**